sexpress/src/main/scala/org/ensime/sexp/SexpParser.scala

```diff
@@ -1,122 +1,166 @@
  * require tweaking, e.g. Scheme's nil, infinity, NaN, etc.
  */
-object SexpParser extends ParboiledParser[Sexp] {
-  protected val Top = SexpP
+object SexpParser {
+
+  def parse(desc: String): Sexp = {
+    val parser = new SexpParser(desc)
+    parser.SexpP.run() match {
+      case Success(d) =>
+        d
+      case Failure(error: ParseError) =>
+        val msg = parser.formatError(error, new ErrorFormatter(showTraces = true))
+        throw new Exception("Failed to parse sexp: " + msg)
+      case Failure(other) =>
+        throw new Exception("Failed to parse sexp: ", other)
+    }
+  }
 
-  // e.g. for .el files
-  def flatParse(el: String): Sexp = parse("(" + el + "\n)")
+  // https://www.gnu.org/software/emacs/manual/html_node/elisp/Basic-Char-Syntax.html
+  // https://www.gnu.org/software/emacs/manual/html_node/elisp/Syntax-for-Strings.html
+  // Not supported: https://www.gnu.org/software/emacs/manual/html_node/elisp/Non_002dASCII-in-Strings.html
+  private[sexp] val specialChars = Map[String, String](
+    "\"" -> "\"",
+    "a" -> 7.toChar.toString,
+    "b" -> "\b",
+    "t" -> "\t",
+    "n" -> "\n",
+    "v" -> 11.toChar.toString,
+    "f" -> "\f",
+    "r" -> "\r",
+    "e" -> 27.toChar.toString,
+    "s" -> " ",
+    "d" -> 127.toChar.toString,
+    "\\" -> "\\"
+  )
+}
 
-  private lazy val SexpP: Rule1[Sexp] = rule("Sexp") {
+/**
+ * Parse Emacs Lisp into an `Sexp`. Other lisp variants may
+ * require tweaking, e.g. Scheme's nil, infinity, NaN, etc.
+ */
+class SexpParser(val input: ParserInput) extends Parser with StringBuilding {
+
+  private def SexpP: Rule1[Sexp] = rule {
     SexpAtomP | SexpListP | SexpEmptyList | SexpConsP | SexpQuotedP
   }
 
-  private lazy val SexpAtomP: Rule1[SexpAtom] = rule("Atom") {
+  private def SexpConsP: Rule1[SexpCons] = rule {
+    LeftBrace ~ SexpP ~ Whitespace ~ "." ~ Whitespace ~ SexpP ~ RightBrace ~> {
+      (x: Sexp, y: Sexp) => SexpCons(x, y)
+    }
+  }
+
+  private def SexpListP: Rule1[Sexp] = rule {
+    LeftBrace ~ SexpP ~ zeroOrMore(Whitespace ~ SexpP) ~ RightBrace ~> {
+      (head: Sexp, tail: Seq[Sexp]) => { SexpList(head :: tail.toList) }
+    }
+  }
+
+  private def SexpAtomP: Rule1[SexpAtom] = rule {
     SexpCharP | SexpStringP | SexpNaNP | SexpNumberP | SexpSymbolP
   }
 
-  private lazy val SexpCharP: Rule1[SexpChar] = rule("Char") {
-    ch('?') ~ NormalChar
-  } ~~> { c => SexpChar(c.head) }
+  private def SexpCharP: Rule1[SexpChar] = rule {
+    '?' ~ NormalChar ~> { (c: String) => SexpChar(c.head) }
+  }
 
-  private lazy val SexpStringP: Rule1[SexpString] = rule("String") {
-    ch('"') ~ zeroOrMore(Character) ~~> (chars => SexpString(chars.mkString(""))) ~ ch('"')
+  private def SexpStringP: Rule1[SexpString] = rule {
+    '"' ~ zeroOrMore(Character) ~ '"' ~> { strs: Seq[String] => SexpString(strs.mkString("")) }
+    //    '"' ~ clearSB() ~ zeroOrMore((Character | "\"\"") ~ appendSB()) ~ '"' ~ push(SexpString(sb.toString))
```

Comment:
What's the problem with the commented out line?
The `mkString` based version above is a lot slower!